File: theory_of_encoding_compression_and_information_recovery/course_work/src/compressor/domain/users/ports/password_hasher.py
from abc import abstractmethod
from typing import Protocol

from compressor.domain.users.values.user_password_hash import UserPasswordHash
from compressor.domain.users.values.user_raw_password import UserRawPassword


class PasswordHasher(Protocol):
    @abstractmethod
    def hash(self, raw_password: UserRawPassword) -> UserPasswordHash: ...

    @abstractmethod
    def verify(self, *, raw_password: UserRawPassword, hashed_password: UserPasswordHash) -> bool: ...

File: сross_platform_software_systems/14/user_service/user_service/src/main/java/com/c3equalz/user_service/domain/user/ports/PasswordHasher.java
package com.c3equalz.user_service.domain.user.ports;

import com.c3equalz.user_service.domain.user.values.RawPassword;
import com.c3equalz.user_service.domain.user.values.UserPasswordHash;

/**
 * Port for password hashing operations.
 * Provides methods to hash raw passwords and verify them against hashed passwords.
 */
public interface PasswordHasher {

    /**
     * Hashes a raw password.
     *
     * @param rawPassword the raw password to hash
     * @return the hashed password
     */
    UserPasswordHash hash(RawPassword rawPassword);

    /**
     * Verifies a raw password against a hashed password.
     *
     * @param rawPassword the raw password to verify
     * @param hashedPassword the hashed password to verify against
     * @return true if the password matches, false otherwise
     */
    boolean verify(RawPassword rawPassword, UserPasswordHash hashedPassword);
}

